File: src/0133-clone-graph/solution.py
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

from typing import Optional

class Solution:
    def cloneGraph(self, node: Optional[Node]) -> Optional[Node]:
        cache = {}
        def dfs(node):
            if node in cache:
                return cache[node]
            newNode = Node(node.val)
            cache[node] = newNode
            for n in node.neighbors:
                cache[node].neighbors.append(dfs(n))
            return newNode
        return dfs(node) if node else None

File: src/0133-clone-graph/solution.ts
/**
 * Definition for _Node.
 * class _Node {
 *     val: number
 *     neighbors: _Node[]
 * 
 *     constructor(val?: number, neighbors?: _Node[]) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.neighbors = (neighbors===undefined ? [] : neighbors)
 *     }
 * }
 * 
 */

const cloned = new Map<Node, Node>();

function cloneGraph(node: Node | null): Node | null {
	if (!node) {
		return null;
	}

	if (cloned.has(node)) {
		return cloned.get(node)!;
	}

	const newNode = new Node(node.val);
	cloned.set(node, newNode);
	newNode.neighbors = node.neighbors
        .map((neighbor) => cloneGraph(neighbor)!);
                                           
	return newNode;
};
